File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function(days, costs) {
  const set = new Set(days);
  const last = days.at(-1);
  const dp = new Array(last + 1).fill(0);
  
  for (let i = 1; i <= last; i++) {
    if (!set.has(i)) dp[i] = dp[i - 1];
    else {
      const [a, b, c] = costs;
      dp[i] = Math.min(
        dp[Math.max(0, i - 1)] + a,
        dp[Math.max(0, i - 7)] + b,
        dp[Math.max(0, i - 30)] + c
      )
    }
  }
  
  return dp.at(-1);
};

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        seen, last = set(days), days[-1]
        dp = [0] * (last + 1)
        for i in range(1, last + 1):
            if i not in seen:
                dp[i] = dp[i - 1]
            else:
                a, b, c = costs
                dp[i] = min(dp[max(0, i - 1)] + a, 
                            dp[max(0, i - 7)] + b, 
                            dp[max(0, i - 30)] + c)
        return dp[-1]
            